src/base/utils/string.h

```diff
@@ -63,10 +65,39 @@ namespace Utils::String
 
     QStringList splitCommand(const QString &command);
 
-    QString join(const QList<QStringView> &strings, QStringView separator);
-
     QString fromDouble(double n, int precision);
 
+    template <typename Container>
+    QString joinIntoString(const Container &container, const QString &separator)
+        requires ExplicitlyConvertibleTo<typename Container::value_type, QString>
+    {
+        auto iter = container.cbegin();
+        const auto end = container.cend();
+        if (iter == end)
+            return {};
+
+        // Some `join` implementation will iterate over the elements to count the total size and other
+        // implementation will not.
+        const qsizetype totalLength = std::accumulate(iter, end, (separator.size() * (container.size() - 1))
+            , [](const qsizetype total, const typename Container::value_type &value)
+        {
+            return total + QString(value).size();
```

Comment:
The one of disadvantages of such generalized implementation is that you cannot be sure that conversion to string is trivial (like just returning inner string). So for some types, such a calculation of the resulting size can negate all the benefits of "reserve".

Reply:
>you cannot be sure that conversion to string is trivial 

Fortunately there are no such drawbacks in our current codebase...
Also if someone really need `joinIntoString()` then they should pay the cost. Otherwise they should use regular `join()`.

>So for some types, such a calculation of the resulting size can negate all the benefits of "reserve".

BTW `reserve()` done correctly potentially results in less memory fragmentation.

Reply:
> > you cannot be sure that conversion to string is trivial
> 
> Fortunately there are no such drawbacks in our current codebase...

Yes. But type-specific implementation can still be even more optimal.

Reply:
I'm sorry, I'm really at a loss to understand why you're trying so hard to promote this general implementation. The only thing that could be considered its advantage is possible code deduplication (if we used it with multiple types, which is not the case). But it's so trivial that it can hardly really be considered significant enough.

Reply:
>Yes. But type-specific implementation can still be even more optimal.

But it won't be usable for other container types (e.g. QSet<QString>, QVarLengthArray, std containers...), isn't? And IIRC you didn't oppose the idea that joining strings is a common operation.

>I'm sorry, I'm really at a loss to understand why you're trying so hard to promote this general implementation. The only thing that could be considered its advantage is possible code deduplication (if we used it with multiple types, which is not the case). 

It is not just potentially 'possible' but it already is. It can replace `QString join(const QList<QStringView> &strings, QStringView separator);` although I'm not sure it is still in use.

Reply:
>The only thing that could be considered its advantage is possible code deduplication 

Not only code deduplication, it also prevents a shallow copy of the container when compared to `QStringList(iter).join()`.

Reply:
> > The only thing that could be considered its advantage is possible code deduplication
> 
> Not only code deduplication, it also prevents a shallow copy of the container when compared to `QStringList(iter).join()`.

I meant the advantage over type-dependent implementations.

Reply:
> But it won't be usable for other container types (e.g. QSet, QVarLengthArray, std containers...), isn't?

Why? The type of container has absolutely nothing to do with it. In "type-specific implementation" I am referring only to the type of value.
Oh well. I'm afraid this discussion is reaching a dead end. If it is enough for you to just avoid creating an intermediate `QStringList` and you want to stop at this optimization, then I don't mind. But at the same time, I reserve for myself the opportunity to improve it further at my leisure.

Reply:
> It can also replace `QString join(const QList<QStringView> &strings, QStringView separator)`

Not really. It may seem strange, but `QStringView` isn't convertible to `QString` (at least in the way `joinIntoString()` expects it).


> although I'm not sure it is still in use.

Yes. Otherwise it won't compile due to stated above.

Reply:
I'll return to the original post:
>So for some types, such a calculation of the resulting size can negate all the benefits of "reserve".

If the problem is at `QString(value)` then one could use compile-time conditional to adapt it. For example:
```c++
[](const qsizetype total, const typename Container::value_type &value)
{
    if constexpr (std::is_same_v<typename Container::value_type, QStringView>)
        return total + value.size();
    else
        return total + QString(value).size();
}
```
and 
```c++
requires ExplicitlyConvertibleTo<typename Container::value_type, QString> || Stringable<typename Container::value_type>
```

Reply:
> If the problem is at `QString(value)` then one could use compile-time conditional to adapt it.

Given that we are not talking about a library function, but about qBittorrent private helper, then of course it could be dependent on some specific types.
`Tag` could also have `size` property, which is quite natural for this type.
Anyway, the above improves only the implementation of `joinIntoString`. But I would still prefer to use a more encapsulated approach for joining tags (although it could use `joinIntoString` internally).

Reply:
> Given that we are not talking about a library function, but about qBittorrent private helper, then of course it could be dependent on some specific types.

But I would prefer to use more abstract/flexible approach like used in standart library. E.g. it can depend only from adapter function template `Utils::String::size(str)` that would return `QString(str).size()` by default and allow types to provide their own specializations for better performance. Then you don't need to touch `joinIntoString()` when any new custom stringable type is added.